### ISSUE TITLE
cluster is very slow

### ISSUE THREAD

**Issue by zhaoxin111** [CASIA-IVA-Lab/ISP-reID#19]

**Comment by zhaoxin111** [comment]
Hi, I try to reproduce the model on market dataset, all the settings are default.
I found the cluster time is very long, ~300s on cpu and 600s on gpu.
The os is linux, and my card is RTX 3090.
Is there any suggestions to improve the cluster?

**Comment by zhukuan96** [comment]
We conduct the clustering every 2 epochs, thus 300s once means 5h for 120epoch. I think it is ok. Sometimes the data IO also affects the times. If you want to accelerate it, you can try to use multithreading to conduct the image clustering in parallel.

**Comment by zhaoxin111** [comment]
> We conduct the clustering every 2 epochs, thus 300s once means 5h for 120epoch. I think it is ok. Sometimes the data IO also affects the times. If you want to accelerate it, you can try to use multithreading to conduct the image clustering in parallel.

Thanks your reply. I had tried to paralle the cluster step by multiprocessing. However，the code was stuck in
![image](https://user-images.githubusercontent.com/9295202/163118087-d8176bcd-830d-45d2-a33a-f624d247754a.png)
![image](https://user-images.githubusercontent.com/9295202/163115949-0333aff0-8946-447c-beda-dc0086a1f1aa.png)

Shoud I replace multiprocessing with multihereading? I think maybe the multi threads in faiss conflicts with current multiprocessing.

**Comment by zhukuan96** [comment]
I donot know, but you can try to turn off the faiss parallelism (if used), because most time is cost on repeated clustering for every ID rather than clustering once.

**Comment by zhaoxin111** [comment]
Thanks for your suggestion, I will try it out.
Besides, I visualized the human body on the occlude dukemtmc query images by your provided trained model, I found that many body regions are missing, but the test metrics on the entire dataset are close to the paper.

![image](https://user-images.githubusercontent.com/9295202/163339054-d40d513b-d094-4f0c-afdf-7dec9ab1f5b0.png)
![image](https://user-images.githubusercontent.com/9295202/163339055-c044e387-ccd7-406f-ae93-37a20ed98346.png)
![image](https://user-images.githubusercontent.com/9295202/163339085-5528b5e6-4c52-4b00-b57d-0eabda6f44d6.png)

**Comment by zhaoxin111** [comment]
Here are some imgs visulized by my model which was trained on Market dataset. The metric is colse to the paper, but the  visualization looks a lot better
![image](https://user-images.githubusercontent.com/9295202/163339547-b54c53c7-0953-4dac-b9f5-732211fab7ab.png)
![image](https://user-images.githubusercontent.com/9295202/163339579-28d86f62-9e84-4c94-812f-a1e16f0890c4.png)
![image](https://user-images.githubusercontent.com/9295202/163339595-7fdc7d07-2f6d-447f-89ef-79758f64aef1.png)

**Comment by zhukuan96** [comment]
Yes, the visualization on holistic dataset is much better, especially on DukeMTMC-reID. You can also adjust the hyper-parameters on line 193, 194  in trainer.py to obtain better visualization.

**Comment by zhaoxin111** [comment]
Since the clustering effect of human body parts on the occlusion dataset is not very good, what is the reason for the high reid accuracy?
As can be seen from the above pictures, only a small part of the human body has been noticed, which means that the local features seem to play a relatively limited role, but the head area seems to be activated.

**Comment by zhaoxin111** [comment]
![image](https://user-images.githubusercontent.com/9295202/163506796-a4a6ea01-9428-4a06-8f94-a968e8b61885.png)
The comparison of learned semantic and extra semantic was done based on holistic dataset，did you compare it on occluded dataset？

**Comment by zhukuan96** [comment]
Sorry, but we have not compared the learned semantic and extra semantic on occluded datasets. Maybe on the occluded dataset, only focusing on the most important parts (or always visiable parts) is enough.

**Comment by zhaoxin111** [comment]
Thank you. I will try to compare them to validate it.
